proposals/0032-middleware-api.md

```diff
@@ -0,0 +1,469 @@
+        locals.validationMessage = "Important information are missing";
+    }
+}
+
+const auth: MiddlewareResponseHandler = async ({ request, redirect }: APIContext, next: MiddlewareNextResponse) => {
+  // The user expectation is that `validation` was already executed (check `sequence`).
+  // This means we don't need to check if `userName` or `password` exit.
+  // If they don't exist, it's an user error.
+  const formData = await request.formData();
+  const userName = formData.get("username");
+  const password = formData.get("password");
+  
+  // We run the authentication using a third-party service
+  const result = await doAuth({ userName, password });
+  if (result.status === "SUCCESS") {
+      return redirect("/secure-area");
+  } else {
+    locals.validationMessage = "User name and/or password are invalid";
+  }
+}
+
+
+export const onRequest = sequence(validation, auth);
+```
+
+It's important to note that `locals` is an object that **lives and dies within a single Astro route**;
+when your route page is rendered, `locals` won't exist anymore and a new one
+will be created.
+
+If a user needs to persist some information that lives among multiple pages
+requests, they will need to store that information somewhere else.
+
+# Testing Strategy
+
+This feature requires integration tests. We need to have tests for  
+multiple scenarios:
+- development server;
+- static build;
+- SSR build;
+- adapters (Node.js, deno, Cloudflare, etc.);
+
+# Drawbacks
+
+The RFC doesn't break any existing code. It's an additional feature that should
+not break the existing behaviour of an Astro application.
+
+Even though the middleware pattern is widely spread among backend frameworks, it's not
+always easy to explain how the pattern works and the user expectations.
+
+I would expect some reports about "why my code doesn't work" and find out that
+the issue was around the user code.
+
+This feature will unlock new patterns inside an Astro application, and I would expect more work from the Documentation Team
+to frame the "most-used recipes" around the usage of middleware.
+
+Due to Astro code base architecture, the implementation of the feature will have to happen in
+three different places, risking having duplicated code or missing logic.
+
+# Alternatives
+
+I have considered implementing middleware using the configuration API.
+
+While this strategy is well-established in the Astro ecosystem, it could slow down
+the implementation of middleware logic in user-land because the user would be forced
+to create some boilerplate just to use the new logic.
+
+While Astro could have provided some API to reduce the boilerplate, this felt tedious. 
+Plus, this would have forced us to **crate another configuration field** where the user
+could specify the order of the middleware.
+
+
+# Adoption strategy
+
+Considering how big the feature is, the API will be released under an experimental flag.
+
+Users can opt in via new flag:
```

Comment:
We usually also have a CLI flag, like `--experimental-middleware`. So you could either use the CLI flag or add it to the Astro config file.